chaste_codegen/cvode_chaste_model.py

```diff
@@ -1,13 +1,60 @@
 import chaste_codegen as cg
+import sympy as sp
 import time
+from chaste_codegen._partial_eval import partial_eval
 
 
 class CvodeChasteModel(cg.ChasteModel):
     """ Holds template and information specific for the CVODE model type"""
-    # TODO:lookup tables
 
     def __init__(self, model, file_name, **kwargs):
         super().__init__(model, file_name, **kwargs)
+        self._use_analytic_jacobian = kwargs.get('use_analytic_jacobian', False)  # store if jacobians are needed
+        self._jacobian_equations, self._jacobian_matrix = self._get_jacobian()
+        self._formatted_state_vars = self._update_state_vars()
+        self._formatted_jacobian_equations, self._formatted_jacobian_matrix_entries = self._format_jacobian()
+
+    def _print_modifiable_parameters(self, symbol):
+        return 'NV_Ith_S(mParameters, ' + str(self._modifiable_parameters.index(symbol)) + ')'
+
+    def _get_jacobian(self):
+        jacobian_equations, jacobian_matrix = [], []
+        if self._use_analytic_jacobian:
+            state_var_matrix = sp.Matrix(self._state_vars)
+            # get deriv eqs and substitute in all variables other than state vars
+            derivative_eqs = partial_eval(self._derivative_equations, self._y_derivatives, keep_multiple_usages=False)
+            # sort by state var so both are in the same order
+            derivative_eqs = sorted(derivative_eqs, key=lambda d: self._state_vars.index(d.lhs.args[0]))
+            # we're only interested in the rhs
+            derivative_eqs = [eq.rhs for eq in derivative_eqs]
+            derivative_eq_matrix = sp.Matrix(derivative_eqs)
+            jacobian_matrix = derivative_eq_matrix.jacobian(state_var_matrix)
+            jacobian_equations, jacobian_matrix = sp.cse(jacobian_matrix, order='none')
+        return jacobian_equations, sp.Matrix(jacobian_matrix)
+
+    def _update_state_vars(self):
+        jacobian_symbols = set()
+        for eq in self._jacobian_equations:
+            jacobian_symbols.update(eq[1].free_symbols)
+        for en in self._jacobian_matrix:
+            jacobian_symbols.update(en.free_symbols)
+        formatted_state_vars = self._formatted_state_vars
+        for i in range(len(formatted_state_vars)):
+            formatted_state_vars[i]['in_jacobian'] = formatted_state_vars[i]['sympy_var'] in jacobian_symbols
+        return formatted_state_vars
+
+    def _format_jacobian(self):
+        assert isinstance(self._jacobian_matrix, sp.Matrix), 'Expecting a jacobian as a matrix'
+        equations = [{'lhs': self._printer.doprint(eq[0]), 'rhs': self._printer.doprint(eq[1])}
+                     for eq in self._jacobian_equations]
+        rows, cols = self._jacobian_matrix.shape
+        jacobian = []
+        for j in range(cols):
+            for i in range(rows):
+                matrix_entry = self._jacobian_matrix.row(i).col(j)[0]
```

Comment:
You can just do `matrix_entry = self._jacobian_matrix[i, j]` for this!

Reply:
aah I didn't know that because length gave me total # of elements I was expecting #row/colum line in other languages :)